### PR TITLE
feat: automatization for the releasing process

### PR DESCRIPTION
**Pull request by josedialpad**
Jira issue: https://switchcomm.atlassian.net/browse/DT-138

## Description
This PR adds [`semantic-release`](https://semantic-release.gitbook.io/semantic-release/) to automatize some steps of our release process including:
- Generating CHANGELOG and Release notes.
- Determining a semantic version bump (based on the types of commits landed).
- Publishing in Github Releases.

`semantic-release` uses the commit messages to determine the consumer impact of changes in the codebase and automatize the things listed above. The convention proposed to adopt is [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).

Added a `commit-msg` hook and Github actions for checking the meeting of this convention on the commits included on the release branches: (`production`, `staging`, `beta` and `alpha`).

Since we're merging PRs using `squash` strategy, a Github action linting the PR title was added to lint over this title since this PR title should be used in the squashed commit message.
Also there is a GH action to lint over merged commits on `staging` and `production` since we cannot reject these commits before pushing (it's a Github Enterprise Server feature only and we're using Github enterprise Cloud).

Forked Dialtone repo to test the process https://github.com/josedialpad/dialtone/.

## Pull Request Checklist

 - [x] Ask the contributors if a similar effort is already in process or has been solved.
 - [x] Review the [contribution guidelines](https://github.com/dialpad/dialtone/blob/staging/.github/CONTRIBUTING.md).
 - [x] Use `staging` as your pull request's base branch. (All PRs using `production` as its base branch will be declined).
 - [x] Ensure all `gulp` scripts successfully compile.
 - [x] Update, remove, or extend all affected documentation.
 - [x] Ensure no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
 - [x] Request a review from Brad Paugh, David Becher, or Drew Chandler.

## Obligatory GIF (super important!)

rocket

### CODE REVIEW

**Comment by josedialpad** [comment]
Well, there is an example of the Lint PR title action already over this PR https://github.com/dialpad/dialtone/runs/3981189904?check_suite_focus=true

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -5,3 +5,21 @@ Thanks for your interest in contributing to Dialtone. Please take a moment to re
 ## Pull Requests
 
 Before submitting a pull request, **please ask first in [a new issue](https://github.com/dialpad/dialtone/issues/new)**. It's always discouraging to have a pull request declined after putting in the time. To avoid this, outline the problem and proposed solution. This includes a rough example of how you might compose the styles.
+
+## Commit message convention
+
+Dialtone use [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/) for the commit messages in the release branches: `production`, `staging`, `alpha` and `beta`. Make sure every commit you want to include in the release notes is meeting this specification.
```
Comment (braddialpad) [inline .github/CONTRIBUTING.md]: Nit: "Dialtone uses"

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -5,3 +5,21 @@ Thanks for your interest in contributing to Dialtone. Please take a moment to re
...
+
+### Git hooks
+
+To enforce your commit on the release branches are correct according to the commit specification, install the `commit-msg` hook.
+
+In the Dialtone root directory run:
+
+```
+npm run prepare
+```
```
Comment (braddialpad) [inline .github/CONTRIBUTING.md]: Users shouldn't have to manually run this. Prepare is a lifecycle script that will automatically execute after `npm install`

File: .github/CONTRIBUTING.md (hunk 1 above)
Comment (braddialpad) [reply .github/CONTRIBUTING.md]: can we also add a commit message example below this?

File: .github/workflows/lint-commit-message.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      - name: Use Node.js
+        uses: actions/setup-node@v1
+        with:
+          node-version: '14.x'
+
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Lint commit message
+        run: echo "${{ github.event.head_commit.message }}" | npx commitlint --verbose
```
Comment (braddialpad) [inline .github/workflows/lint-commit-message.yml:24]: Looks like we just output failures right now, fine for now but in the future maybe we could have slack or email notifications to the user

**Review by braddialpad — CHANGES_REQUESTED** [review]
Couple small things. Thanks Jose this looks great! Really excited to see this in action

File: .github/CONTRIBUTING.md (hunk 1 above)
Comment (josedialpad) [reply .github/CONTRIBUTING.md]: Thanks, fixed the typo https://github.com/dialpad/dialtone/pull/484/commits/4a99be5f153456ef8eea72f748e7536c8e738233 and added examples and more details in a `COMMIT_CONVENTION` file https://github.com/dialpad/dialtone/pull/484/commits/8289d55c24e91431a041310fd88664b3a886c25c.

File: .github/CONTRIBUTING.md (hunk 2 above)
Comment (josedialpad) [reply .github/CONTRIBUTING.md]: Good catch! It was just part of the installation the first time necessary to run it manually but then is as you said. Updated https://github.com/dialpad/dialtone/pull/484/commits/d05e23a89aeb25175def1e63f6af6d6e7654ce19

**Review by braddialpad — APPROVED** [review]
Great thanks!

File: .github/COMMIT_CONVENTION.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
+## Git Commit Message Convention
+
+Dialtone uses [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/) for the commit messages in the release branches: `production`, `staging`, `alpha` and `beta`. Make sure every commit you want to include in the [semantic release process](RELEASING.md) is meeting this specification. When opening a PR, ensure the PR's title also complies with the Conventional Commits specification. This way, if your PR is merged using squash strategy, the PR's title will be used af the commit message.
```
Comment (ryandialpad) [inline .github/COMMIT_CONVENTION.md]:
```suggestion
Dialtone uses [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/) for the commit messages in the release branches: `production`, `staging`, `alpha` and `beta`. Make sure every commit you want to include in the [semantic release process](RELEASING.md) is meeting this specification. When opening a PR, ensure the PR's title also complies with the Conventional Commits specification. This way, if your PR is merged using squash strategy, the PR's title will be used as the commit message.
```

Looks like there is a small typo on this line

**Review by ryandialpad — APPROVED** [review]
w/ comment

Great work!

File: .github/workflows/lint-commit-message.yml (hunk 3 above)
Comment (dbecherdialpad) [reply .github/workflows/lint-commit-message.yml:24]: so if you mess up the commit message, how do you fix it? at this point it's already been pushed to origin

Comment (braddialpad) [reply .github/workflows/lint-commit-message.yml:24]: You would have to amend it directly on staging. Would have preferred a pre-receive hook but unfortunately it's not supported on Github. Hopefully it wouldn't happen often as we also have the pr title check.

Comment (josedialpad) [reply .github/workflows/lint-commit-message.yml:24]: Another safest way will be to revert that commit and push a new commit with the correct message.

**Pull request merged**